rust_src/src/multibyte.rs

```diff
@@ -348,13 +357,214 @@ impl LispStringRef {
         s.size
     }
 
-    /// Return width of STRING when displayed in the current buffer. Width is
-    /// measured by how many columns it occupies on the screen. When calculating
-    /// width of a multibyte character in STRING, only the base leading-code is
-    /// considered; the validity of the following bytes is not checked.  Tabs in
-    /// STRING are always taken to occupy `tab-width' columns.
+    /// Return the byte index corresponding to the given character index.
+    ///
+    /// Caches the ouput of the most recent calculation to make several calls on
+    /// the same string faster.
+    pub fn char_idx_to_byte(self, char_index: usize) -> usize {
+        let mut best_below = 0;
+        let mut best_below_byte = 0;
+        let mut best_above = self.len_chars() as usize;
+        let mut best_above_byte = self.len_bytes() as usize;
+
+        // Return if contents are unibyte.
+        if best_above == best_above_byte {
+            return char_index;
+        }
+
+        // Use cache operating on the cached string?
+        unsafe {
+            if LispObject::from(self) == string_index_cache {
+                // Check if character is before or after cached position
+                if charpos_cache < char_index {
+                    best_below = charpos_cache;
+                    best_below_byte = bytepos_cache;
+                } else {
+                    best_above = charpos_cache;
+                    best_above_byte = bytepos_cache
+                }
+            }
+        }
+
+        let byte_index = if char_index - best_below < best_above - char_index {
+            // Target is forwards.
+            let mut byte = best_below_byte;
+
+            while best_below < char_index {
+                byte += self.head_len(byte);
+                best_below += 1;
+            }
+
+            byte
+        } else {
+            // Target is backwards.
+            let mut byte = best_above_byte;
+
+            while best_above > char_index {
+                byte -= 1;
+                while !self.is_head_at(byte) {
+                    byte -= 1;
```

Comment:
Why is it backwards one but forwards `head_len(byte)`?

Reply:
The difference is that there's two while loops. When moving backwards, you can't use the head length to preemptively check how big a jump you need to make, so you move backwards byte by byte until you hit a head.